bezier-rs/lib/src/utils.rs

```diff
@@ -64,3 +66,138 @@ pub fn solve_quadratic(discriminant: f64, two_times_a: f64, b: f64, c: f64) -> V
 	}
 	roots
 }
+
+/// Compute the cube root of a number.
+fn cube_root(f: f64) -> f64 {
+	if f < 0. {
+		-(-f).powf(1. / 3.)
+	} else {
+		f.powf(1. / 3.)
+	}
+}
+
+/// Solve a cubic of the form `x^3 + px + q`, derivation from: <https://trans4mind.com/personal_development/mathematics/polynomials/cubicAlgebra.htm>.
+pub fn solve_reformatted_cubic(discriminant: f64, a: f64, p: f64, q: f64) -> Vec<f64> {
+	let mut roots = Vec::new();
+	if p == 0. {
+		roots.push(cube_root(-q));
+	} else if q == 0. {
+		if p < 0. {
+			roots.push((-p).powf(1. / 2.));
+		}
+	} else if discriminant == 0. {
+		let q_divided_by_2 = q / 2.;
+		let a_divided_by_3 = a / 3.;
+		// all roots are real, and 2 are repeated
+		roots.push(2. * cube_root(-q_divided_by_2) - a_divided_by_3);
+		roots.push(cube_root(q_divided_by_2) - a_divided_by_3);
+	} else if discriminant > 0. {
+		// one real and two imaginary roots
+		let q_divided_by_2 = q / 2.;
+		let square_root_discriminant = discriminant.powf(1. / 2.);
+		roots.push(cube_root(-q_divided_by_2 + square_root_discriminant) - cube_root(q_divided_by_2 + square_root_discriminant) - a / 3.);
+	} else {
+		// three real roots
+		let p_divided_by_3 = p / 3.;
+		let a_divided_by_3 = a / 3.;
+		let cube_root_r = (-p_divided_by_3).powf(1. / 2.);
+		let phi = (-q / (2. * cube_root_r.powi(3))).acos();
+
+		let two_times_cube_root_r = 2. * cube_root_r;
+		// three real roots
+		roots.push(two_times_cube_root_r * (phi / 3.).cos() - a_divided_by_3);
+		roots.push(two_times_cube_root_r * ((phi + 2. * PI) / 3.).cos() - a_divided_by_3);
+		roots.push(two_times_cube_root_r * ((phi + 4. * PI) / 3.).cos() - a_divided_by_3);
+	}
+	roots
+}
+
+/// Solve a cubic of the form `ax^3 + bx^2 + ct + d`.
+pub fn solve_cubic(a: f64, b: f64, c: f64, d: f64) -> Vec<f64> {
+	if a.abs() <= 1e-5 {
+		if b.abs() <= 1e-5 {
+			// if both a and b are approximately 0, treat as a linear problem
+			solve_linear(c, d)
+		} else {
+			// if a is approximately 0, treat as a quadratic problem
+			let discriminant = c * c - 4. * b * d;
+			solve_quadratic(discriminant, 2. * b, c, d)
+		}
+	} else {
+		let new_a = b / a;
+		let new_b = c / a;
+		let new_c = d / a;
+
+		// Refactor cubic to be of the form: a(t^3 + pt + q), derivation from: https://trans4mind.com/personal_development/mathematics/polynomials/cubicAlgebra.htm
+		let p = (3. * new_b - new_a * new_a) / 3.;
+		let q = (2. * new_a.powi(3) - 9. * new_a * new_b + 27. * new_c) / 27.;
+		let discriminant = (p / 3.).powi(3) + (q / 2.).powi(2);
+		solve_reformatted_cubic(discriminant, new_a, p, q)
+	}
+}
+
+/// Compare two `f64` numbers with a provided max absolute value difference.
+pub fn compare_f64(f1: f64, f2: f64, max_abs_diff: f64) -> bool {
```

Comment:
would it make more sense to call this f64_approximately_equals to be consistent with the in range functions?